### PR TITLE
Add rr commands to remove and move traces

### PR DESCRIPTION
**Pull request by vinhill**
As discussed in #1415, this pull request implements `rr rm` and `rr mv` commands to manage traces.

See #3558 for a similar pull request using `std::filesystem` that was closed because that library caused build errors.

### CODE REVIEW

File: src/RmCommand.cc [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    default:
+      assert(0 && "Unknown option");
+  }
+
+  return true;
+}
+
+static int _rm_cb(const char* path, const struct stat*, int, struct FTW*) {
+  int ret = remove(path);
+  if (ret)
```
Comment (rocallahan) [inline src/RmCommand.cc:60]: {} around if body

File: src/TraceStream.h [hunk 2]
```diff
@@ -523,6 +523,7 @@ class TraceReader : public TraceStream {
 
 extern std::string trace_save_dir();
 extern std::string resolve_trace_name(const std::string& trace_name);
+extern std::string latest_trace_symlink();
```
Comment (rocallahan) [inline src/TraceStream.h:526]: These don't actually need `extern` do they? let's remove it

File: src/util.cc [hunk 3]
```diff
@@ -2192,6 +2192,64 @@ bool is_directory(const char* path) {
...
+  }
+  return true;
+}
+
+bool is_valid_trace_name(const string& entry, bool log_error) {
+  // filename corresponds to dirname
+  const string name = filename(entry.c_str());
+
+  if (name.empty()) {
+    if (log_error) fprintf(stderr, "\nrr: Trace name cannot be empty\n\n");
```
Comment (rocallahan) [inline src/util.cc]: These all need {} around the if body, and `fprintf` on its own line

**Review by rocallahan — CHANGES_REQUESTED** [review]

File: src/MvCommand.cc [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    to_path = trace_save_dir() + "/" + to;
+  }
+
+  string to_fname = filename(to_path.c_str());
+  if (to_fname == "latest-trace") {
+    fprintf(stderr, "\nrr: Cannot rename to latest-trace.\n\n");
+    return 1;
+  }
+
+  if (from_path == to_path) {
```
[the remark this answers is not in the document]
Comment (vinhill) [reply src/MvCommand.cc:60]: I check that `to_path` doesn't exist and `real_path` requires an existing path.
But given `from_path` exists,  I guess checking `from_path==to_path` is irrelevant.

File: src/util.h [hunk 5]
```diff
@@ -515,6 +515,36 @@ ssize_t pwrite_all_fallible(int fd, const void* buf, size_t size, off64_t offset
...
+ * Deletes the latest_trace symlink, logs an error and returns false on failure.
+ */
+bool remove_latest_trace_symlink();
+
+/*
+ * Returns whether |entry| is a valid trace name and optionally logs why not.
+ * I.e. does not start with . or #, does not end with ~, is neither cpu_lock
+ * nor latest_trace.
+ */
+bool is_valid_trace_name(const std::string& entry, bool log_error=false);
```
[the remark this answers is not in the document]
Comment (vinhill) [reply src/util.h]: Yeah, that function isn't pretty. If I understand you correctly, `is_valid_trace_name(entry, TraceNameCheck::verbose)` also doesn't seem best to me. Is a `string*` out param okay?

File: src/MvCommand.cc [hunk 6]
```diff
@@ -90,21 +88,21 @@ static int mv(const string& from, const string& to, FILE* out) {
     }
   }
 
-  int ret = rename(from_path.c_str(), to_path.c_str());
-
-  if (ret) {
+  if (int ret = rename(from_path.c_str(), to_path.c_str())) {
```
Comment (rocallahan) [inline src/MvCommand.cc:91]: Remove the declaration of `ret` here, it's causing a build error on CI.

**Review by rocallahan — CHANGES_REQUESTED** [review]
The rest looks fine.

**Comment by rocallahan** [comment]
Thanks!

**Pull request merged**